url86119f019a_solzy_hive_parse_tgz-pJ8-org_apache_hadoop_hive_ql_parse_MainJ8/src/org/apache/hadoop/hive/ql/parse/HiveParser_IdentifiersParser.java

```diff
@@ -0,0 +1,12423 @@
+
+				adaptor.addChild(root_0, root_1);
+				}
+
+			}
+
+			else // 178:12: -> {$dist == null}? ^( TOK_FUNCTION functionName ( ( selectExpression )+ )? ( $ws)? )
+			if (dist == null) {
+				// IdentifiersParser.g:178:32: ^( TOK_FUNCTION functionName ( ( selectExpression )+ )? ( $ws)? )
+				{
+				CommonTree root_1 = (CommonTree)adaptor.nil();
+				root_1 = (CommonTree)adaptor.becomeRoot((CommonTree)adaptor.create(TOK_FUNCTION, "TOK_FUNCTION"), root_1);
+				adaptor.addChild(root_1, stream_functionName.nextTree());
+				// IdentifiersParser.g:178:60: ( ( selectExpression )+ )?
+				if ( stream_selectExpression.hasNext() ) {
+					if ( !(stream_selectExpression.hasNext()) ) {
+						throw new RewriteEarlyExitException();
+					}
+					while ( stream_selectExpression.hasNext() ) {
+						adaptor.addChild(root_1, stream_selectExpression.nextTree());
+					}
+					stream_selectExpression.reset();
+
+				}
+				stream_selectExpression.reset();
+
+				// IdentifiersParser.g:178:82: ( $ws)?
+				if ( stream_ws.hasNext() ) {
+					adaptor.addChild(root_1, stream_ws.nextTree());
+				}
+				stream_ws.reset();
+
+				adaptor.addChild(root_0, root_1);
+				}
+
+			}
+
+			else // 179:29: -> ^( TOK_FUNCTIONDI functionName ( ( selectExpression )+ )? )
+			{
+				// IdentifiersParser.g:179:32: ^( TOK_FUNCTIONDI functionName ( ( selectExpression )+ )? )
+				{
+				CommonTree root_1 = (CommonTree)adaptor.nil();
+				root_1 = (CommonTree)adaptor.becomeRoot((CommonTree)adaptor.create(TOK_FUNCTIONDI, "TOK_FUNCTIONDI"), root_1);
+				adaptor.addChild(root_1, stream_functionName.nextTree());
+				// IdentifiersParser.g:179:62: ( ( selectExpression )+ )?
+				if ( stream_selectExpression.hasNext() ) {
+					if ( !(stream_selectExpression.hasNext()) ) {
+						throw new RewriteEarlyExitException();
+					}
+					while ( stream_selectExpression.hasNext() ) {
+						adaptor.addChild(root_1, stream_selectExpression.nextTree());
+					}
+					stream_selectExpression.reset();
+
+				}
+				stream_selectExpression.reset();
+
+				adaptor.addChild(root_0, root_1);
+				}
+
+			}
+
+
+			retval.tree = root_0;
+			}
+
+			}
+
+			retval.stop = input.LT(-1);
+
+			if ( state.backtracking==0 ) {
+			retval.tree = (CommonTree)adaptor.rulePostProcessing(root_0);
+			adaptor.setTokenBoundaries(retval.tree, retval.start, retval.stop);
+			}
+			if ( state.backtracking==0 ) { gParent.msgs.pop(); }
+		}
+
+		catch (RecognitionException e) {
+		  throw e;
+		}
+
+		finally {
+			// do for sure before leaving
+		}
+		return retval;
+	}
+	// $ANTLR end "function"
+
+
+	public static class functionName_return extends ParserRuleReturnScope {
+		CommonTree tree;
+		@Override
+		public CommonTree getTree() { return tree; }
+	};
+
+
+	// $ANTLR start "functionName"
+	// IdentifiersParser.g:182:1: functionName : ( KW_IF | KW_ARRAY | KW_MAP | KW_STRUCT | KW_UNIONTYPE | identifier );
+	public final HiveParser_IdentifiersParser.functionName_return functionName() throws RecognitionException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 277 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.